### PR TITLE
feat: Add KMDF Sample Driver with pool leak to demonstrate Driver Verifier

### PR DESCRIPTION
**Pull request by svasista-ms**
The `pool_leak` sample KMDF driver contains an intentional error that demonstrates the capabilities and features of [Driver Verifier](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/driver-verifier) and the [Device Fundamental tests](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/device-fundamentals-tests).
    
The driver uses WDM ExAllocatePool2 API to allocate memory to its Device Context buffer and intentionally does not free this memory anywhere in the driver. Such driver behavior can lead to vulnerabilities in the system if not caught. By enabling the Driver Verifier on this driver, the pool leak violation can be caught and with an active KDNET session, the bug can be examined and analyzed further.

This PR adds the sample driver to the workspace with instructions in the project's README to build, deploy and test the driver. The sample also demonstrates the use of [WDTF](https://learn.microsoft.com/en-us/windows-hardware/drivers/wdtf/wdtf-overview), [TAEF ](https://learn.microsoft.com/en-us/windows-hardware/drivers/taef/getting-startedr) and the Device Fundamental Tests to test a driver.

### CODE REVIEW

File: tools/dv/samples/kmdf/pool_leak/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation.
+// License: MIT OR Apache-2.0
+
+//! # Abstract
+//!
+//!     This KMDF sample contains an intentional error that is designed to
```
Comment (NateD-MSFT) [inline tools/dv/samples/kmdf/pool_leak/src/lib.rs:6]: These extra spaces at the start of lines cause the doc to render incorrectly.  Please remove them.

```suggestion
//! This KMDF sample contains an intentional error that is designed to
```

File: tools/dv/samples/kmdf/pool_leak/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//! demonstrate the capabilities and features of Driver Verifier and the Device
+//! Fundamental tests.
+//!     
+//!     The driver is desgined to allocate memory using ExAllocatePool2 to its
```
Comment (NateD-MSFT) [inline tools/dv/samples/kmdf/pool_leak/src/lib.rs:10]:
```suggestion
//! The driver is designed to allocate memory using ExAllocatePool2 to its
```

File: tools/dv/samples/kmdf/pool_leak/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+//! Device Context buffer when a device is added by the PnP manager. However,
+//! this buffer is not freed anywhere in the driver, including the driver unload
+//! function.
+//!
+//!     By enabling the Driver Verifier on this driver, the pool leak
```
Comment (NateD-MSFT) [inline tools/dv/samples/kmdf/pool_leak/src/lib.rs:15]:
```suggestion
//! By enabling the Driver Verifier on this driver, the pool leak
```

File: tools/dv/samples/kmdf/pool_leak/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+//! violation can be caught when the driver is unloaded and with an active KDNET
+//! session, the bug can be analyzed further.
+
+#![no_std]
+#![cfg_attr(feature = "nightly", feature(hint_must_use))]
+#![deny(clippy::all)]
+#![warn(clippy::pedantic)]
+#![warn(clippy::nursery)]
+#![warn(clippy::cargo)]
+#![allow(clippy::missing_safety_doc)]
```
Comment (NateD-MSFT) [inline tools/dv/samples/kmdf/pool_leak/src/lib.rs:25]: Clippy complains about CamelCase entries in docs when they don't have backticks around them (i.e. `MyFunction`.)  You _should_ go and add backticks to function names in your documentation name, but we need to add a suppression here to stop it from complaining about "PnP".

```suggestion
#![allow(clippy::missing_safety_doc)]
#![allow(clippy::doc_markdown)]
```

File: tools/dv/samples/kmdf/pool_leak/src/driver.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    // Get the device context and initialize it. WdfObjectGet_DEVICE_CONTEXT is an
+    // inline function generated by WDF_DECLARE_CONTEXT_TYPE macro in the
+    // device.h header file. This function will do the type checking and return
+    // the device context. If you pass a wrong object  handle
+    // it will return NULL and assert if run under framework verifier mode.
+    let device_context = unsafe { wdf_object_get_device_context(device as WDFOBJECT) };
+    unsafe {
+        let length: usize = 64;
```
Comment (NateD-MSFT) [inline tools/dv/samples/kmdf/pool_leak/src/driver.rs:156]: This 64 seems like a magic number.  Where did it come from?

**Review by NateD-MSFT — CHANGES_REQUESTED** [review]

File: tools/dv/samples/kmdf/pool_leak/src/driver.rs (hunk 5 above)
Comment (svasista-ms) [reply tools/dv/samples/kmdf/pool_leak/src/driver.rs:156]: I arbitrarily chose 64 bytes as the number of bytes to be allocated by ExAllocatePool2

Comment (NateD-MSFT) [reply tools/dv/samples/kmdf/pool_leak/src/driver.rs]: OK.  I might make a comment at this line that this is the memory we intend to leak, and then update the comment in evt_driver_unload to note that this line is where the leaked allocation happened, just to make it easier for people to read.

**Review by NateD-MSFT — APPROVED** [review]

**Comment by NateD-MSFT** [comment]
@wmmc88 Can you take a look at the dependency review pipeline here?

**Comment by wmmc88** [comment]
> @wmmc88 Can you take a look at the dependency review pipeline here?

Of the top of my head, the github dependency review action has several issues with AND and OR licenses, and evaluating them correctly. In the past, I've had to work around those issues. See comments in https://github.com/microsoft/windows-drivers-rs/blob/9d4ab1a1a8c14587e8354aaadfb83ea49beebc05/.github/workflows/github-dependency-review.yaml#L19 and https://github.com/microsoft/Windows-rust-driver-samples/blob/64c426eb7f21b9f46946084039e2b68be30f59bc/.github/workflows/github-dependency-review.yaml#L19.

I do believe there was some recent upstream change in the action. I'll see if i have some time to look into this later this week.

**Comment by wmmc88** [comment]
> > @wmmc88 Can you take a look at the dependency review pipeline here?
> 
> Of the top of my head, the github dependency review action has several issues with AND and OR licenses, and evaluating them correctly. In the past, I've had to work around those issues. See comments in https://github.com/microsoft/windows-drivers-rs/blob/9d4ab1a1a8c14587e8354aaadfb83ea49beebc05/.github/workflows/github-dependency-review.yaml#L19 and https://github.com/microsoft/Windows-rust-driver-samples/blob/64c426eb7f21b9f46946084039e2b68be30f59bc/.github/workflows/github-dependency-review.yaml.
> 
> I do believe there was some recent upstream change in the action. I'll see if i have some time to look into this later this week.

@svasista-ms try cherry picking this commit onto your branch: https://github.com/wmmc88/Windows-rust-driver-samples/pull/2/commits/871632dbc035bd7558f7d9d6802c176b7474e23a

It looks like the accepted format for "allow-licenses" was changed to be a single string that's comma delimited. the yaml `>-` syntax does that. I've made the same change in `windows-drivers-rs` repo already

**Review by wmmc88 — APPROVED** [review]

**Review by hamzamust — APPROVED** [review]

**Pull request merged**